File: java/Graph.java
import java.util.ArrayList;
import java.util.TreeMap;
import java.util.Map;
import java.util.Collections;

public class Graph{
    // New ArrayList to store the names of the vertices.
    private ArrayList<String> vertices = new ArrayList<String>();

    // New ArrayList to store TreeMaps containing the edges.
    private ArrayList<TreeMap<String,Integer>> edges
	= new ArrayList<TreeMap<String,Integer>>();

    // New int to track the number of vertices.
    private int numVerts = 0;

    // New int to track the number of edges.
    // Even though the addEdge function adds the edge in two locations
    // (to make the graph undirected), this keeps track of the single
    // edge that was added.
    private int numEdges = 0;

    // New int to track the total weight of the graph.
    private int totWeight = 0;

    public void addEdge(String vert1, String vert2, int weight){
	// Get the index location of each vertex to pull the edge data.
	int loc1 = vertices.indexOf(vert1);
	int loc2 = vertices.indexOf(vert2);
	
	// Unpack the TreeMap for each vertex of the edge.
	TreeMap<String,Integer> adjList1 = edges.get(loc1);
	TreeMap<String,Integer> adjList2 = edges.get(loc2);
	
	// Add the edge to the list.
	adjList1.put(vert2,weight);
	adjList2.put(vert1,weight);
	numEdges++;
	totWeight += weight; // Update weight.
	
	// Save the updated TreeMap.
	edges.set(loc1,adjList1);
	edges.set(loc2,adjList2);
    }
    
    public void addVert(String newVert){
	if(!vertices.contains(newVert)){
	    // Add the new vertex if it's not already in the list.
	    vertices.add(newVert);
	    numVerts++;

	    // Sort the vertex list lexicographically.
	    Collections.sort(vertices);
	    
	    // Add a new TreeMap that corresponds to this vert.
	    edges.add(vertices.indexOf(newVert), new TreeMap<String,Integer>());
	}
    }

    public ArrayList<String> getEdges(){
	// Create an ArrayList to store the edges formatted for printing.
	ArrayList<String> formattedEdges = new ArrayList<String>();

	for(int i = 0; i < edges.size(); i++){
	    // Unpack to create the ArrayList.
	    TreeMap<String,Integer> edgeMap = edges.get(i);

	    // Loop through every edge in the TreeMap that was just unpacked.
	    for (Map.Entry<String,Integer> entry : edgeMap.entrySet()) {
		String formattedEdge;
		String reverseFormattedEdge;

		// Create string of the edge in proper format.
		formattedEdge =  "(" + vertices.get(i) + ",";
		formattedEdge += entry.getKey() + "," + entry.getValue() + ")";
		
		// Create a string of the same edge, but vertices swapped.
		// Will use this to make sure the edge is not in the array
		// already.
		reverseFormattedEdge =  "(" + entry.getKey() + ",";
		reverseFormattedEdge += vertices.get(i) + ","
		    + entry.getValue() + ")";
				
		// Check if reverseFormattedEdge is in ArrayList.
		// Add formattedEdge if not.
		if(!formattedEdges.contains(reverseFormattedEdge)){
		    formattedEdges.add(formattedEdge);
		}
	    }
	}
	return formattedEdges;
      }

    public TreeMap<String,Integer> getEdges(String vert){
	int location = vertices.indexOf(vert);

	return edges.get(location);
    }

    public String getMinEdge(int loc){
	// Pull the edges for a certain vertex.
	TreeMap<String,Integer> edgeSet = edges.get(loc);

	// Keep track of the key for the smallest edge.
	String minKey = edgeSet.firstKey(); 

	// Loop through the edges, finding the smallest edge.
	for (Map.Entry<String,Integer> entry : edgeSet.entrySet()) {
	    // Update the tracker var if a smaller edge is found.
	    if(entry.getValue() < edgeSet.get(minKey)){
		minKey = entry.getKey();
	    }
	}
	return minKey;
    }

    public ArrayList<String> getVerts(){
	return vertices;
    }

    public String getVerts(int loc){
	return vertices.get(loc);
    }

    /* This function was added for debugging purposes.
    public void printEdges(){
	System.out.println("Edges:");
	for(int i = 0; i < edges.size(); i++){
	    System.out.print(vertices.get(i) + "\t-> ");
	    // Unpack for printing.
	    TreeMap<String,Integer> temp = edges.get(i);
	    for (Map.Entry<String,Integer> entry : temp.entrySet()) {
		System.out.print("(" + entry.getKey() + ","
				 + entry.getValue()+ ")");
	    }
	    System.out.print("\n");
	}
     }*/

    /* This function was added for debugging purposes.
    public void printVerts(){
	System.out.println("Vertices:");
	for(int i = 0; i < vertices.size(); i++){
	    System.out.print(i+1);
	    System.out.println(": " + vertices.get(i));
	}
     }*/

    public void removeEdge(String vert1, String vert2){
	// Get the index location of each vertex to pull the edge data.
	int loc1 = vertices.indexOf(vert1);
	int loc2 = vertices.indexOf(vert2);

	// Unpack the TreeMap for each vertex of the edge.
	TreeMap<String,Integer> adjList1 = edges.get(loc1);
	TreeMap<String,Integer> adjList2 = edges.get(loc2);

	// Remove the edge from the list.
	totWeight -= adjList1.get(vert2); // Update weight.
	adjList1.remove(vert2);
	adjList2.remove(vert1);
	numEdges--; // Update the number of edges in the graph.
	
	// Save the updated TreeMap.
	edges.set(loc1,adjList1);
	edges.set(loc2,adjList2);
    }

    public int sizeEdges(){
	return numEdges;
    }

    public int sizeVerts(){
	return numVerts;
    }

    public int weight(){
	return totWeight;
    }
}

File: c/graph.c
/**
 * @file graph.c
 * @author Carter Crews
 * @date  28 Apr 2017
 * @brief Implementation file for functionality over graphs.
 *
 * This file contains the implementation of the functions that operate over graphs.
 *
 */

#include "graph.h"

void addEdge(char vert1[VERT_NAME_LEN], char vert2[VERT_NAME_LEN],
	     char weight[VERT_NAME_LEN],
	     char vertexNames[NUM_VERTICES][VERT_NAME_LEN], int* vertexSize,
	     char edges[NUM_VERTICES][NUM_EDGES][2][VERT_NAME_LEN],
	     int edgeSize[NUM_VERTICES])
{
  int vert1Loc = findVertIndex(vert1,*vertexSize,vertexNames);
  int vert2Loc = findVertIndex(vert2,*vertexSize,vertexNames);

  /* If a vertex we are adding isn't in the array yet, add it. */
  if(vert1Loc == -1)
    {
      strcpy(vertexNames[*vertexSize],vert1);
      vert1Loc = *vertexSize;
      (*vertexSize)++;
    }
  if(vert2Loc == -1)
    {
      strcpy(vertexNames[*vertexSize],vert2);
      vert2Loc = *vertexSize;
      (*vertexSize)++;
    }

  /* Add the edge to both vertices to ensure graph is undirected. */
  strcpy(edges[vert1Loc][edgeSize[vert1Loc]][0],vert2);
  strcpy(edges[vert1Loc][edgeSize[vert1Loc]][1],weight);
  edgeSize[vert1Loc]++;
  
  strcpy(edges[vert2Loc][edgeSize[vert2Loc]][0],vert1);
  strcpy(edges[vert2Loc][edgeSize[vert2Loc]][1],weight);
  edgeSize[vert2Loc]++;

  return;
}

int findVertIndex(char findMe[], int vertexSize,
		  char vertexNames[NUM_VERTICES][VERT_NAME_LEN])
{
  int index = -1;
  int i = 0; 

  while(i < NUM_VERTICES && index == -1)
    {
      if(strcmp(findMe, vertexNames[i]) == 0)
	{
	  index = i;
	}
      i++;
    }

  return index;
}

int getMinEdge(int vert, char edges[NUM_VERTICES][NUM_EDGES][2][VERT_NAME_LEN],
	       int edgeSize[NUM_VERTICES])
{
  int minEdgeLoc = -1;
  int minEdgeWeight = -1;
  int i;

  for(i = 0; i < edgeSize[vert]; i++)
    {
      /* Make sure the location we're looking at actually contains an edge. */
      if(atoi(edges[vert][i][1]) != -1)
	{
	  /* If we've found a smaller weight, or if this is the first edge,
	     update tracking variables. */
	  if(atoi(edges[vert][i][1]) < minEdgeWeight || minEdgeWeight == -1)
	    {
	      minEdgeLoc = i;
	      minEdgeWeight = atoi(edges[vert][i][1]);
	    }
	}
    }

  return minEdgeLoc;
}

void preventCycles(char vertexNames[NUM_VERTICES][VERT_NAME_LEN],
		   int vertexSize,
		   char edges[NUM_VERTICES][NUM_EDGES][2][VERT_NAME_LEN],
		   int edgeSize[NUM_VERTICES],
		   char treeVertexNames[NUM_VERTICES][VERT_NAME_LEN],
		   int treeVertexSize)
{
  int i,j;

  for(i = 0; i < vertexSize; i++)
    {
      for(j = 0; j < edgeSize[i]; j++)
	{
	  int vert1Loc;
	  int vert2Loc;
	  char vert1[VERT_NAME_LEN];
	  char vert2[VERT_NAME_LEN];
	  char weight[VERT_NAME_LEN];

	  strcpy(vert1,vertexNames[i]);
	  strcpy(vert2,edges[i][j][0]);
	  strcpy(weight,edges[i][j][1]);

	  if(strcmp(vert2,"\0") != 0)
	    {
	      vert1Loc = findVertIndex(vert1,treeVertexSize,treeVertexNames);
	      vert2Loc = findVertIndex(vert2,treeVertexSize,treeVertexNames);

	      if(vert1Loc != -1 && vert2Loc != -1)
		{
		  removeEdge(vert1,vert2,weight,vertexNames,vertexSize,edges,
			     edgeSize);
		}
	    }
	}
    }
}

void printGraph(char vertexNames[NUM_VERTICES][VERT_NAME_LEN], int vertexSize,
		char edges[NUM_VERTICES][NUM_EDGES][2][VERT_NAME_LEN],
		int edgeSize[NUM_VERTICES])
{
  int i,j;

  for(i=0; i<vertexSize; i++)
    {
      for(j=0; j<edgeSize[i]; j++)
	{
	  if(strcmp(vertexNames[i],edges[i][j][0]) < 0)
	    {
	      printf("(%s,%s,%s)\n",
		     vertexNames[i],edges[i][j][0],edges[i][j][1]);
	    }
	}
    }
}

void readData(char vertexNames[NUM_VERTICES][VERT_NAME_LEN], int* vertexSize, 
	      char edges[NUM_VERTICES][NUM_EDGES][2][VERT_NAME_LEN],
	      int edgeSize[NUM_VERTICES], FILE* file)
{
  char tempVertex1[VERT_NAME_LEN];
  char tempVertex2[VERT_NAME_LEN];
  char weight[VERT_NAME_LEN];
  int vert1Index = -1;
  int vert2Index = -1;

  /* Prime the while loop with an initial read. */
  fscanf(file,"%s",tempVertex1);
  fscanf(file,"%s",tempVertex2);
  fscanf(file,"%s",weight);
  vert1Index = findVertIndex(tempVertex1,*vertexSize,vertexNames);
  vert2Index = findVertIndex(tempVertex2,*vertexSize,vertexNames);

  /* While loop to read in vertices. */
  while(!feof(file))
    {
      /* Check if vertices exist in memory. If not, add them. */
      if(vert1Index == -1)
	{
	  strcpy(vertexNames[*vertexSize],tempVertex1);
	  vert1Index = *vertexSize;
	  (*vertexSize)++;
	}
      if(vert2Index == -1)
	{
	  strcpy(vertexNames[*vertexSize],tempVertex2);
	  vert2Index = *vertexSize;
	  (*vertexSize)++;
	}

      /* Copy the data read in to the array of edges. */
      strcpy(edges[vert1Index][edgeSize[vert1Index]][0],tempVertex2);
      strcpy(edges[vert1Index][edgeSize[vert1Index]][1],weight);
      edgeSize[vert1Index]++;

      /* Make the graph undirected. */
      strcpy(edges[vert2Index][edgeSize[vert2Index]][0],tempVertex1);
      strcpy(edges[vert2Index][edgeSize[vert2Index]][1],weight);
      edgeSize[vert2Index]++;

      /* Next read from file. */
      fscanf(file,"%s",tempVertex1);
      fscanf(file,"%s",tempVertex2);
      fscanf(file,"%s",weight);
      vert1Index = findVertIndex(tempVertex1,*vertexSize,vertexNames);
      vert2Index = findVertIndex(tempVertex2,*vertexSize,vertexNames);
    }

  return;
}

void removeEdge(char vert1[VERT_NAME_LEN], char vert2[VERT_NAME_LEN],
		char weight[VERT_NAME_LEN],
		char vertexNames[NUM_VERTICES][VERT_NAME_LEN], int vertexSize,
		char edges[NUM_VERTICES][NUM_EDGES][2][VERT_NAME_LEN],
		int edgeSize[NUM_VERTICES])
{
  int vert1Loc = findVertIndex(vert1,vertexSize,vertexNames);
  int vert2Loc = findVertIndex(vert2,vertexSize,vertexNames);
  int vert1Edge,vert2Edge;
  int i;

  /* Find the location of the edges in the array. */
  for(i = 0; i < edgeSize[vert1Loc]; i++)
    {
      if(strcmp(vert2,edges[vert1Loc][i][0]) == 0)
	{
	  vert1Edge = i;
	}
    }
  for(i = 0; i < edgeSize[vert2Loc]; i++)
    {
      if(strcmp(vert1,edges[vert2Loc][i][0]) == 0)
	{
	  vert2Edge = i;
	}
    }

  /* Remove the edges. 
     Update the weight with -1 to indicate that no edge exists in this 
     location. */
  strcpy(edges[vert1Loc][vert1Edge][0],"\0");
  strcpy(edges[vert1Loc][vert1Edge][1],"-1");
  
  strcpy(edges[vert2Loc][vert2Edge][0],"\0");
  strcpy(edges[vert2Loc][vert2Edge][1],"-1");
  
  return;
}
